main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala

```diff
@@ -0,0 +1,358 @@
+//
+//    object Body {
+//
+//      case class Atom(pred: Name.Pred, den: Denotation, polarity: Ast.Polarity, terms: List[ErasedAst.Term.Body], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Predicate.Body
+//
+//      case class Guard(exp: ErasedAst.Expression[_], terms: List[ErasedAst.Term.Body], loc: SourceLocation) extends ErasedAst.Predicate.Body
+//
+//    }
+//
+//  }
+//
+//  object Term {
+//
+//    sealed trait Head {
+//      def tpe: MonoType
+//    }
+//
+//    object Head {
+//
+//      case class QuantVar(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Head
+//
+//      case class CapturedVar(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Head
+//
+//      case class Lit(sym: Symbol.DefnSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Head
+//
+//      case class App(exp: ErasedAst.Expression[_], args: List[Symbol.VarSym], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Head
+//
+//    }
+//
+//    sealed trait Body {
+//      def tpe: MonoType
+//    }
+//
+//    object Body {
+//
+//      case class Wild(tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Body
+//
+//      case class QuantVar(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Body
+//
+//      case class CapturedVar(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Body
+//
+//      case class Lit(sym: Symbol.DefnSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Body
+//
+//    }
+//
+//  }
+//
+//  case class Attribute(name: String, tpe: MonoType)
+//
+//  case class Case(sym: Symbol.EnumSym, tag: Name.Tag, tpeDeprecated: MonoType, loc: SourceLocation)
+//
+//  case class CatchRule(sym: Symbol.VarSym, clazz: java.lang.Class[_], exp: ErasedAst.Expression[_])
+//
+//  case class Constraint(cparams: List[ConstraintParam], head: Predicate.Head, body: List[Predicate.Body], loc: SourceLocation)
+//
+//  sealed trait ConstraintParam {
+//    def sym: Symbol.VarSym
+//  }
+//
+//  object ConstraintParam {
+//
+//    case class HeadParam(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.ConstraintParam
+//
+//    case class RuleParam(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.ConstraintParam
+//
+//  }
+//
+//  case class FormalParam(sym: Symbol.VarSym, tpe: MonoType)
+//
+//  case class FreeVar(sym: Symbol.VarSym, tpe: MonoType)
+
+  sealed trait Expression[+T <: JType] {
+    def loc: SourceLocation
+  }
+
+  sealed trait JType
+
+  object JType {
+
+    sealed trait JByte8 extends JType
+
+    sealed trait JShort16 extends JType
+
+    sealed trait JInt32 extends JType
+
+    sealed trait JLong64 extends JType
+
+    sealed trait JChar extends JType
+
+    sealed trait JFloat32 extends JType
+
+    sealed trait JDouble64 extends JType
+
+    sealed trait JBool extends JType
+
+    sealed trait JObject extends JType
```

Comment:
In some sense, we can start where we only have the above types + the boxed variants.

Then we can erase everything to JObject (that is not e.g. an integer).

Then, over time, we can start using precise types like e.g. JRef below.

Does that make sense?